rasa/core/interpreter.py

```diff
@@ -297,6 +302,80 @@ def _load_interpreter(self) -> None:
         self.interpreter = Interpreter.load(self.model_directory)
 
 
+def find_same(action_name, attribute, training_data):
+    for ex in training_data.training_examples:
+        if (
+            ex.as_dict()["text"] == action_name
+            and ex.get(SPARSE_FEATURE_NAMES[attribute]) is not None
+        ):
+            example = ex
+            break
+    return example
+
+
+class RasaE2EInterpreter(NaturalLanguageInterpreter):
+    def __init__(self,):
+        self._load_interpreter()
+
+    def parse(self, text: Text):
+        result = self.interpreter.parse(text, only_output_properties=False)
+        return result
+
+    def _load_interpreter(self) -> None:
+        from rasa.nlu import config
+
+        config = config.load()
+        self.trainer = Trainer(config)
+        self.interpreter = None
+
+    def prepare_training_data(self, trackers_as_states, trackers_as_actions, domain):
+        training_examples = []
+        for tracker in trackers_as_states:
+            for state in tracker:
+                if state and not state == {}:
+                    training_examples.append(state["prev_action"])
+                    if "user" in state.keys():
+                        training_examples.append(state["user"])
+        training_examples += [Message(action) for action in domain.action_names]
+        training_data = TrainingData(training_examples=training_examples)
+
+        training_examples = [action for tr in trackers_as_actions for action in tr]
+        training_data.training_examples += training_examples
+        return training_data
+
+    def prepare_training_data_and_train(
+        self, trackers_as_states, trackers_as_actions, output_path, domain
+    ):
+        """
+        Trains the vertorizers from real data;
+        Preliminary for when the featurizer is to be used for raw text;
+        Args:
+             - trackers_as_states: real data as a dictionary
+             - trackers_as_actions: output label
+        """
+        training_data = self.prepare_training_data(
+            trackers_as_states, trackers_as_actions, domain
+        )
+        kwargs = {'no_copy': True}
+        self.interpreter = self.trainer.train(training_data, **kwargs)
+
+        ## Not so clever fix for now;
+        ## TODO: properly encode everything in place;
+        for tracker in trackers_as_states:
+            for state in tracker:
```

Comment:
@evgeniiaraz Not relevant for the architecture, but general coding comments: Avoid more than two / three levels of indentation. Use things like guard clauses (https://refactoring.guru/replace-nested-conditional-with-guard-clauses) or extract the content of a `for`/`if` clause as separate function.